ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/RaceConditionIT.scala

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.testtool.suites
+
+import com.daml.ledger.api.testtool.infrastructure.Allocation._
+import com.daml.ledger.api.testtool.infrastructure.Assertions._
+import com.daml.ledger.api.testtool.infrastructure.LedgerTestSuite
+import com.daml.ledger.api.v1.transaction.{TransactionTree, TreeEvent}
+import com.daml.ledger.api.v1.value.RecordField
+import com.daml.ledger.test.semantic.RaceTests._
+import com.daml.lf.data.{Bytes, Ref}
+
+import scala.concurrent.Future
+import scala.util.Success
+
+final class RaceConditionIT extends LedgerTestSuite {
+
+  test(
+    "WWDoubleNonTransientCreate",
+    "Cannot concurrently create multiple non-transient contracts with the same key",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, alice)) =>
+    val Attempts = 100
+    val ExpectedNumberOfSuccessfulCreations = 1
+    Future
+      .traverse(1 to Attempts) { case _ =>
+        ledger.create(alice, ContractWithKey(alice)).transform(Success(_))
+      }
+      .map { results =>
+        assertLength(
+          "Successful contract creations",
+          ExpectedNumberOfSuccessfulCreations,
+          results.filter(_.isSuccess),
+        )
+        ()
+      }
+  })
+
+  test(
+    "WWDoubleArchive",
+    "Cannot archive the same contract multiple times",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, alice)) =>
+    val Attempts = 100
+    val ExpectedNumberOfSuccessfulArchivals = 1
+    for {
+      contract <- ledger.create(alice, ContractWithKey(alice))
+      _ <- Future.traverse(1 to Attempts) { _ =>
+        ledger.exercise(alice, contract.exerciseContractWithKey_Archive).transform(Success(_))
+      }
+      transactions <- ledger.transactionTrees(alice)
+    } yield {
+      import TransactionUtil._
+      assertLength(
+        "Successful contract archivals",
+        ExpectedNumberOfSuccessfulArchivals,
+        transactions.filter(isArchival),
+      )
+      ()
+    }
+  })
+
+  test(
+    "WWArchiveVsNonTransientCreate",
+    "Cannot re-create a contract if it hasn't yet been archived",
```

Comment:
I believe here you're referring to **_re-assign_** a contract key. Can you amend the description?

Reply:
Done.